### DISCUSSION TITLE
Populating the views.

### DISCUSSION THREAD

**Discussion by meteos77** [textbrowser/biblioteq#118] [Q&A] [answered]
I tested a base provided in the forum 
discussion https://github.com/textbrowser/biblioteq/discussions/105
the database has 1000 books
- I connect the database and nothing is displayed although the category is on books ?
I reapply on books and a message "populating the views" appears ... but it takes 12 seconds on my machine to fill 

question is this normal or did I make a mistake ?
with 1000 books it's 12 seconds on a good machine

knowing that the goal is to manage 10000 books on a very average machine, I will have to wait 2 minutes "populating the views 

I tried to add a book and it redoes "populating the views" on the 1000 books

what did I misunderstand ?
thank you in advance

**Comment by textbrowser** [comment]
See image.

![Screenshot_20211107_080942](https://user-images.githubusercontent.com/10701156/140646291-f038ce13-14e7-4f5c-8951-fd5cf2d46f29.png)

Enable pagination. For SQLite, set a reasonable number of items per page. Also disable Populate View on Item Creation if you're viewing an unlimited number of items.

**Comment by textbrowser** [comment]
The delay is also caused by the progress dialog. Perhaps that should be optional. Now, if you're loading 1,000 items, you may desire to interrupt the process; hence the progress dialog. Also remember that it is not wise to display a large number of items as they can deplete the resources on the machine.

**Reply by meteos77** [reply]
I compare with other software
tellico for example loads the 11700 lines in about 2 seconds
in 2 seconds I have neither the desire to interrupt the process nor the interest to do.


I removed all the options: same problem
track to explore why it takes so long on biblioteq ?

- for comparison on your machine 10000 articles is loaded in how long? 
- effectively remove the counter : I see the number of pages vibrate
- remove the thumbnails

**Reply by textbrowser** [reply]
Yes, the images may be the cause. Did you set the pagination? Remember, BQ is intended for large collections. One would not browse the entire NYC library at once. You would paginate some section. You'd search and paginate the results. Both PQ and SQLite have hard and soft limits, hence the pagination. BQ does not and cannot display 1,000,000 rows in a few seconds. I can see about making the images optional.

**Reply by textbrowser** [reply]
On PQ over WiFi, 50 items are displayed in approximately 1.5 seconds. This is caused by the progress dialog and the images. I will test without the two and measure the elapsed time rigorously and I know it will be "instantaneous."

**Reply by textbrowser** [reply]
100 books loaded from PQ over WiFi without progress dialogs and images: 323 milliseconds.

**Reply by textbrowser** [reply]
Conclusion: the progress dialog and the images are the cause.
The workaround: smaller pages.
The solution: optional images. Now, this is difficult because it requires all queries to become more complicated. The images reside in the databases. I did not design BQ to load 10,000 items on a single display. :)
Anyway, have to do other things now. Will work on optional images and progress dialogs later.

**Reply by textbrowser** [reply]
116 milliseconds: 100 books over WiFi from PQ. This test removed additional image processing.

**Reply by textbrowser** [reply]
A new idea also. Pagination will be available with the Members Browser.

**Comment by meteos77** [comment]
thank you for looking at my problem

knowing that I have no cover image nor back cover in the base it's only text, the option without image will suit me perfectly for the moment

once found the document by the search that the photo is displayed on the individual record is very cool

but who searches in a base of 10000 entries by the images ?

**Comment by textbrowser** [comment]
An image search does not exist. The images are included with items.

**Comment by textbrowser** [comment]
When I search for an item, I visually review the images because I find it easier to discover the item that I'm interested in than by reading titles.

**Comment by meteos77** [comment]
I want to make it clear that I am not criticizing biblioteQ at all.
I am a user who looks for features and compares. 
In return for your great work, I inform you about possible improvements.
this is the minimum for a user involved in Free Software. I translate the docs and soon the menus (still some fine tuning to do).
thanks again for your software

I don't know how to send you a message about a competitor?

**Comment by textbrowser** [comment]
Criticism has a weird way of becoming judgement and judgement is negative; and this word varies by culture, region, person. However, criticism is also a form of critique, evaluation, analysis. So, we can have severe faultfinding or evaluation. People can work towards a solution or not. A project can satisfy a person while another person is impossible to satisfy.

I understand criticisms and critiques and I can usually see when a criticism is insulting. One must be able to critique ideas, products, ideologies, whatever. These evaluations are valuable. Some improve an idea while others reconstruct it. It's complicated. There are probably studies on the merits of critiques, etc.

So, I do not mind ideas, critiques, evaluations. They'll be considered then and if not then, they'll be recorded to be considered later.

A competitor? I don't consider BQ as competing with others. It's something I enjoy and if I can help others enjoy it, I work towards that. Of course, the requests have to be evaluated. :) So, you can post ideas here because it's transparent.

**Reply by meteos77** [reply]
I wrote this preamble because my message talking about a competitor did not arrive at its destination? probably a bad management of my action. sorry

on the forum, since the beginning, there is only you and me ... so I wonder about my approach 
I can think that you developer: who is this user who always ask for new stuff and pollute the forum?

I don't forget that you do this as a hobby.

as you say on the criticism, competition is a word that varies according to the culture, the region, the person. 
even translations can be problematic:  

when I speak about competitor it is from my point of view
competitor between 2 free softwares to choose the one that will best fit our needs.

is there a word in english to express this concept better than competitor ?

to finish I prefer to be transparent in my approach, it's my way to practice my hobby.

Sincerely yours

Translated with www.DeepL.com/Translator (free version)

**Comment by textbrowser** [comment]
It is beautiful to be an underdog. No one knows you exist while those who do may enjoy your presence. :) The forum is empty so there's nothing to pollute.

That's exactly your choice. Which best suits your needs. As you're offering ideas, I will implement those ideas even if you select something else. Grey Literature was born from a request where the person decided to adopt something else.

**Comment by meteos77** [comment]
Hello, you are a bit too much in the stars for my automatic translator :-)

thanks for your modifications it's great to have the result in < 10 sec

idea : there is still the dialog box, which is useless because no info in it


result of your modifications choice of "show tables images

version of 2021/11/07 => loading of the table view 45 seconds for 7032 results
version of 2021/11/08 => loading of the table view 8,5 seconds for 7032 results with option images checked
version of 2021/11/08 => loading of the table view 8,5 seconds for 7032 results without the option images checked

it's really the counter that was slowing down 

comparison : 
tellico 2,8 seconds for 11772 thumbnails
tellico 3,8 seconds for 11772 table lines
I don't know how it works in his code but it goes super fast.

Translated with www.DeepL.com/Translator (free version)

**Comment by textbrowser** [comment]
It's both. If you have large images, which can happen, both the loading of the images from the local (or non-local) database and the progress dialog (which has a habit of repainting) can degrade performance.

I'm currently making the images optional (an option already exists). If images are disabled, the progress dialogs are as well. Of course, the Icon View will be bland.

**Comment by textbrowser** [comment]
Also remember that BQ is works on Raspberry Pis, Sun Ultra 60s, the ancient Apple machines. The purpose of the progress dialogs is to allow the process to be interrupted. And may I mention it again? BQ is not designed to display 100,000 items in a single view. We can work towards the unlimited view in SQLite, but remember that that idea implies that your system is capable of viewing infinitely-many items. I prefer pagination because catalogs typically are partitioned into sections and those sections are partitioned and perhaps those paritions are partitioned. We can discuss the merits of pagination separately. From my perspective, BQ provides both solutions: paginated or single views. The progress dialogs have purpose and I understand that they ought to be optional; which they shall be. In Android, for example, the system provides something called Adapters and AdapterViews. In that system, the items which are visible exist in memory. As one scrolls, items which are invisible are discarded and new items are loaded into memory. That's an interesting framework because it conserves resources. Qt does not have such a system although I implemented something similar to Android's Adapter for another project; a smaller portion of Android's capability. Right, so infinite scrolling (or viewing) is restricted both by the resources and the the sizeof(int) (you need to be able to refer to a row and a size). With pagination, limits are expanded. I prefer pagination because of stated and unstated reasons.

**Comment by textbrowser** [comment]
Android is not limitless with these views because item positions use integers. :)

**Comment by textbrowser** [comment]
Alright, I've modified all of the queries to support optional images. All of the progress dialogs related to populating the main table are now optional too. Test and let me know the results. If you encounter particular queries which are problematic, include the category in the reply. Thanks!

**Comment by textbrowser** [comment]
New option!

![Screenshot_20211108_214308](https://user-images.githubusercontent.com/10701156/140852438-9136f50c-f527-47b7-bc50-0066bdc76b39.png)

**Comment by textbrowser** [comment]
Did you test the new changes?

* Disable images.
* Disable progress dialogs.

If so, are you satisfied with the results?

**Comment by meteos77** [comment]
you did a fantastic job
the population time statistics in the table show it
thanks again for adding the options  I like the table tool tips.


_reminder suggestion if very very easy to implement otherwise I don't ask anything (it's a little extra)
have a keyboard shortcut to interrupt the BiblioteQ Progress Dialog process: xxx
the Cancel button with an F8 shortcut because Ctrl+C is already taken

in "show progress dialogs" mode deactivated
this would allow to interrupt without disturbing the speed ?

reminder suggestion if very very easy to implement otherwise I don't ask anything (it's a little extra)_

**Comment by textbrowser** [comment]
Let me add a tool tip describing the intent of Show Progress Dialogs. The option disables of enables progress dialogs whenever loading of main-table data occurs. The option is situated in the Main Window page of the Other Options. The option is enabled by default.

Permettez-moi d'ajouter une info-bulle décrivant l'intention de Show Progress Dialogs. L'option désactive ou active les boîtes de dialogue de progression chaque fois que le chargement des données de la table principale se produit. L'option est située dans la page de la fenêtre principale des autres options. L'option est activée par défaut.

**Comment by textbrowser** [comment]
Shortcut will be added. :) If I can get access to the button widget.

Un raccourci sera ajouté. :) Si je peux accéder au widget bouton.

**Comment by meteos77** [comment]
I dream or what! wonderful you speak French
you have a magic tool that does everything by itself ?
je rêve ou quoi ! merveilleux vous parlez français
vous avez un outils magique qui fait tout tous seul ?

**Comment by textbrowser** [comment]
Non, GH n'a rien de tel. Je sais comment jouer avec le système.

**Comment by meteos77** [comment]
I really appreciate the effort you made to make a message in French! 
thank you very much

**Comment by textbrowser** [comment]
Vous êtes donc satisfait de l'amélioration des performances. Le plus excellent. Cette question est maintenant répondue.

**Reply by meteos77** [reply]
very very satisfied 
I did a test on another machine 
Aspire One Atom570 1,66GHz 3325,35 SSD M4-CT064SSD2 64Go with the version: 2021-10-30 it took for my test: 143 seconds

I will see with the new version

**Comment by meteos77** [comment]
yes it is

**Comment by textbrowser** [comment]
Comme je le soupçonnais, je dois découvrir le widget du bouton de fermeture et lui attribuer un raccourci car QProgressDialog ne propose pas son propre widget. Heureusement, je peux également lui fournir mon propre widget qui aura le raccourci F8. Je n'ai pas trouvé de sens standard de F8 ; donc je vais vous croire sur parole. L'info-bulle dans Autres options détaillera également le raccourci.

**Comment by meteos77** [comment]
what is GH ?

**Comment by textbrowser** [comment]
GitHub. :P

**Reply by meteos77** [reply]
GitHub (/ɡɪt hʌb/; operated as GitHub, Inc.) is a web-based software development hosting and management service, using Git version control software

it's been 15 days since I came across the c++ ... some understanding for beginners

**Comment by textbrowser** [comment, marked as answer]
Completed!

https://github.com/textbrowser/biblioteq/commit/18603107f3869f4765a1fd689fd80caf872825d8

**Comment by meteos77** [comment]
you want us to face the dialog with english / french
in each message, that allows to learn the 2
vous voullez que l'on face le dialogue avec anglais / français
dans chaque message , cela permet d'apprendre les

**Comment by textbrowser** [comment]
Je sais lire entre les lignes. :)

**Comment by meteos77** [comment]
back you are lucky between the lines of c++ :)
my compilation on pi, I thought it would last long enough but no everything is OK

**Comment by meteos77** [comment]
the F8 key works but that's not what I had in mind
I can't express myself well
for me the F8 key was used to interrupt when the panel was not present

in "display progress dialogs" mode disabled
it would allow to interrupt without disturbing the speed ?

but let's not disperse :-)
if making the messages in French makes you lose time, don't bother, I translate for me
I prefer that you do : english -> c++ :-)

**Comment by textbrowser** [comment]
Tirez-moi dessus ! Interrompre quel panneau ?

Les boîtes de dialogue de progression sur la table/vue principale sont désormais facultatives. Supposons que vous chargez 1000 articles et que vous souhaitiez interrompre le chargement. F8 fait ça. Quel autre panneau voulez-vous interrompre ?

Les boîtes de dialogue de progression doivent être redessinées afin que cette nouvelle peinture entrave le processus de remplissage. On ne peut pas tout avoir. :)

**Reply by meteos77** [reply]
cette nouvelle peinture entrave le processus de remplissage.  non-French sentence

**Comment by meteos77** [comment]
when there is the sign it is OK 
but without the sign nothing

quand il ya le panneau c'est OK 
mais sans le panneau rien


proposal label button with "Interrupt or F8"  ou  Ctrl+I si vous préferez rester dans les lettres (if you prefer to stay in the letters )

**Comment by meteos77** [comment]
do you want me to do some tests?

**Comment by textbrowser** [comment]
I would like a description of which process you'd like to interrupt. You cannot interrupt the loading of the main page unless there is a means to interrupt it. F8 cannot interrupt a process unless another process listens to that event. Hence, the progress dialogs. :) You can kill the main process but that's not interrupting it; that's destroying it. Anyway, a precise description. For me, the loading of 10,000 items immediately is a strange concept because I would not scroll over 10,000 items. Now, 10,000 data points being visually represented is something else. A catalog system is partitioned. We wish to read some sections, not 10,000 of them. That is the purpose of the progress dialogs. That's also the reason of pagination. So, I created an option where dialogs are optional. Another option disables or enables item images. What is the end desire?

**Comment by meteos77** [comment]
super satisfied, do not add anything

**Comment by textbrowser** [comment]
Are you sure? I am not fixed on amending it. If you need more preferences, I can add preferences.
